### ISSUE TITLE
add a color swatch to the color adjuster widgets

### ISSUE THREAD

**Issue by jeisner** [mypaint/mypaint#491]
The various color adjuster widgets let you adjust the color by clicking or dragging.
A swatch that tracks the currently selected color is helpfully shown in the lower-left corner of the MyPaint window.
However, that's rather far away from the widget in many cases, and it's hard to look back and forth.

**So it would be nice if the widget itself also showed such a swatch.**

_Where should the swatch go?_ One possibility: instead of redesigning each widget's layout to accommodate a swatch, one possible placement would be to use the currently selected color as the background of the widget's title bar.  The title text would have to be shown in a contrasting color (perhaps always black or white).

_But how about widgets in the panel?_  When a widget is shown in the panel instead of as a popup, there is no title bar per se.  In this case, perhaps you could still use the cutout area between the tab handle and the settings wrench?  This is where the title bar would ordinarily be.  To avoid an overly busy display on the panel, I'd suggest coloring that cutout area only while the pointer is over that particular widget -- and not when the user is drawing or manipulating the color via some other widget.)

_Remark:_ The background chooser color selector does show its own swatch (see #469), but it's the only one that does so at present.  And arguably, it needs the swatch least, since the _whole background_ is changing as the user drags the color.  (Incidentally, the background chooser is also the only place where the HSV rotating triangle widget can still be found.  That's a very nice widget so I wonder why it's no longer available to select drawing colors?)

**Comment by achadwick** [comment]
This decision was taken to simplify the adjuster appearances, and to allow them to be packed into a common container (like the popups) while retaining their natural shapes. The swatch was moved next to the picker widget in the bottom left, which can be clicked to pup up a menu of all colour selectors nearby. This should suit your needs.

I simply don't know what you mean by "the background of the widget's title bar".

I might accept patches which do overlays for new new colour a bit like the colour picker mode's overlay.

Do you have any mockups?

GtkHSV has been deprecated for a very long time, and has been removed from our colour adjuster hierarchy accordingly (#268). It is a very buggy, old widget (#169), and has never been a good fit in the Mypaint UI (#266). If you miss it and are willing to contribute to its maintenance and rehabilitation, you need to collaborate with the GTK+ maintainers, not us <https://bugzilla.gnome.org/>.

**Comment by achadwick** [comment]
**Triage note**: needs UI mockups, or a patch. The functionality is (IMO) already present. Close as wontfix after a week if none of the needed information is forthcoming.

**Comment by jeisner** [comment]
> The functionality is (IMO) already present.

The goal is to get the swatch and the widget into the **same field of view**.   As noted in the original post, yes the functionality is there, but is placed in the lower-left corner, which may be far away from the popup.  Consider for example the liquid wash or the component sliders -- it's hard to navigate to nearby colors unless you're looking at the widget itself, but without a swatch in the same field of view, it's also hard to know what's wrong with the current color.

> I simply don't know what you mean by "the background of the widget's title bar".
> ... needs UI mockups ...
![swatch-in-adjuster](https://cloud.githubusercontent.com/assets/2395181/11131754/39d024d4-895a-11e5-9bfc-b4f10e6cbbfc.png)

> GtkHSV has been deprecated for a very long time ... .It is a very buggy, old widget

Got it.  +1 for your suggestion on @266 of replicating that design within MyPaint eventually.

**Comment by 0ion9** [comment]
+1 on that, now that I understand it. I would describe it as 'swatch color in popup header'  personally.
Would it be too confusing to show the old and new colors as two halves of the popup header background? I think that would give the best accuracy of comparison.

**Comment by jeisner** [comment]
> Would it be too confusing to show the old and new colors as two halves of the popup header background? I think that would give the best accuracy of comparison

@0ion9 Do you mean comparing to the last drawing color used?  In other words, the same comparison that's being done in the lower left corner of MyPaint?  

 It might indeed be more useful.  On the other hand, it makes the layout less pretty / more cluttered / harder to parse.  So I wonder whether the comparison in the lower left is enough to support this.  

What you're saying is that there's really a 3-way comparison: (1) the last drawing color, (2) the currently selected color, and (3) other selectable colors on the widget that one might choose.  So the question is whether it's adequate to handle that as two 2-way comparisons -- (1) vs. (2) in the lower left, and (2) vs. (3) in the widget -- or whether the widget really ought to show all three.  

Showing 1 vs. 2 colors in the popup header can't have very different implementations, so it might be sensible to try it both ways and ask users for their opinion ... or try it one way and see how users react.

**Comment by achadwick** [comment]
Not keen on putting anything in the slack space after the tabs. There may be other tabs between the selector and the options/close button, or the space may be completely filled with tabs and overflow arrows. It's also technically tricky (although not impossible I think) to draw to that particular area. Furthermore, the tabbed notebook widgets are (by convention) not addressable from a colour selector packed within them: they belong to the workspace implementation, and should not be touched by their inhabitants.

The popup colour selector proposal is neater. There are a couple of questions it poses:

1. How would the preview work for the brush chooser popup?
2. How to avoid duplication of comparators above and below when the colour chooser popup is popped up from the status bar colour comparator widget?

Both proposals suffer in that the widgets they draw to are generic widgets unrelated to colour selection, which makes overriding what they do a) that bit more tricky, and b) a source of visual inconsistency within the sidebar, and between popups as noted above.

I would really favour a comparator drawn within the selectors if we can, keeping the design minimal. This may mean making the comparators transient, visible only around times when the program sees motion or colour selection events.

@0ion9 @jeisner What do you think of my suggestion to use something like the picker mode's overlay on colour selectors? Too distracting to have anything following the mouse?

**Comment by jeisner** [comment]
> Not keen on putting anything in the slack space after the tabs.

I like it visually (at least as a temporary thing that appears when you're interacting with the widget).  But another option for dockables would be to increase the height enough to accommodate a similar horizontal swatch at the top of the window's main area, just below the slack space.  This would be reasonably visually consistent with the design I sketched for popups.

> There may be other tabs between the selector and the options/close button, or the space may be completely filled with tabs and overflow arrows.

Yes, I figured the swatch would just be resized to fit the available width, and eliminated if too narrow to be useful (rare?).

> How would the preview work for the brush chooser popup?

No preview there, I think.  Are you thinking of previewing a canonical brush stroke under the current settings?  Seems tricky to define what's wanted, and it would be hard to read the header (e.g., "Set #1").  A preview isn't really needed anyway -- that's the point of the brush icons, many of which include a drawing of a stroke.  But the color selector has many more choices on display, and most of them are only a single pixel - that's why a larger preview is needed.

> How to avoid duplication of comparators above and below when the colour chooser popup is popped up from the status bar colour comparator widget?

I wondered this too.  Options:
1. Duplicate and live with it.  
2. Suppress the preview in the widget in this case.  (Similar to my notion that if several color widgets are active, you don't want all of them previewing.)
3. Just eliminate the status bar widget, under @0ion9's idea that a full comparator appears in the color selector anyway. 

> What do you think of my suggestion to use something like the picker mode's overlay on colour selectors? Too distracting to have anything following the mouse?

It's a cute design.  Certainly puts the information very near where you're looking, and consistent with the color picker.

The overlay obscures the selector a bit, but importantly doesn't obscure the area right around the pointer, since it's a little distance away.  A possible disadvantage is that it keeps moving -- I might find a stationary swatch to be less distracting and easier to evaluate.  

But overall I'd be happy with your version.  I'm on the fence about whether or not it's preferable to the one I sketched.  But as you say, my version poses some implementation challenges, especially for dockables.

(The concern about duplicating the status bar comparator shows up in both designs.  So does @0ion9's question about whether the widget should actually display a comparator and not just a swatch.)
